### PR TITLE
Add secrets config test of replacing TLS certificate

### LINKED ISSUE #199: Add secrets config test to replace the TLS certificate
This issue is to add a new test suite to verify the functionality of snapped `secrets-config` CLI. The steps for such test can be based on the documentation: https://docs.edgexfoundry.org/3.0/getting-started/Ch-GettingStartedSnapUsers/#changing-tls-certificates

Some testing code can be borrowed from what was removed in https://github.com/canonical/edgex-snap-testing/pull/188

For simplicity, the TLS cert creation can be done in bash, similar to https://gist.github.com/farshidtz/8fe88373a5ef1243847282f29a06184f

### PR DESCRIPTION
**Pull request by MonicaisHer**
This PR will resolve #199, and bases on https://github.com/canonical/edgex-snap-testing/pull/202.

TODO:
- [x] After the merge of #202, this PR needs to be rebased to the main branch

### CODE REVIEW

File: test/suites/edgexfoundry/secrets_config_test.go [hunk 1]
```diff
@@ -41,3 +43,26 @@ func TestAddAPIGatewayUser(t *testing.T) {
...
+func TestChangeTlsCert(t *testing.T) {
+	t.Log("Create TLS certificate")
+	createTlsCert(t)
+
+	const coreDataEndpoint = "https://localhost:8443/core-data/api/v2/ping"
+	t.Log("Calling API gateway using new TLS certificates:", coreDataEndpoint)
+	// Note: %%	is a literal percent sign
+	code, _, _ := utils.Exec(t, fmt.Sprintf("curl --show-error --silent --include"+
+		"--output /dev/null --write-out '%%{http_code}' --caCert ca.cert '%s'",
+		coreDataEndpoint))
```
Comment (farshidtz) [inline test/suites/edgexfoundry/secrets_config_test.go:58]: This code uses a file that was created in a script, far away from here, which can change without knowing it is used here. I wonder if we can make this less fragile.

Some comments would be helpful.

**Review by farshidtz — CHANGES_REQUESTED** [review]

**Comment by farshidtz** [comment]
See also:
```
=== RUN   TestChangeTlsCert
    secrets_config_test.go:50: Create TLS certificate
    exec.go:19: [exec] ../../utils/create-tls-certificates.sh
    secrets_config_test.go:54: Calling API gateway using new TLS certificates: https://localhost:8443/core-data/api/v2/ping
    exec.go:19: [exec] curl --show-error --silent --include--output /dev/null --write-out '%{http_code}' --caCert ca.cert 'https://localhost:8443/core-data/api/v2/ping'
    exec.go:72: [stderr] curl: option --include--output: is unknown
        curl: try 'curl --help' or 'curl --manual' for more information
    exec.go:75: exit status 2
```

**Review by farshidtz — COMMENTED** [review]
See also:
```
=== RUN   TestChangeTlsCert
    secrets_config_test.go:50: Create TLS certificate
    exec.go:19: [exec] ../../utils/create-tls-certificates.sh
    secrets_config_test.go:54: Calling API gateway using new TLS certificates: https://localhost:8443/core-data/api/v2/ping
    exec.go:19: [exec] curl --show-error --silent --include--output /dev/null --write-out '%{http_code}' --caCert ca.cert 'https://localhost:8443/core-data/api/v2/ping'
    exec.go:72: [stderr] curl: option --include--output: is unknown
        curl: try 'curl --help' or 'curl --manual' for more information
    exec.go:75: exit status 2
```

File: test/suites/edgexfoundry/secrets_config_test.go [hunk 2]
```diff
@@ -41,3 +44,28 @@ func TestAddAPIGatewayUser(t *testing.T) {
...
+	const caCertFile = "./ca.cert"
+	// Note: %%	is a literal percent sign
+	// Note: The path to the ca.cert file is created by create-tls-certificates.sh
+	code, _, _ := utils.Exec(t, fmt.Sprintf(
+		"curl --show-error --silent --include --output /dev/null --write-out '%%{http_code}' --caCert %s '%s'",
+		caCertFile,
+		coreDataPingEndpoint))
+
+	// A success response should return status 401 because the endpoint is protected.
+	require.Equal(t, "401", strings.TrimSpace(code))
```
Comment (MonicaisHer) [inline test/suites/edgexfoundry/secrets_config_test.go:64]: Got different result locally and on Github Action:
local:
```
exec.go:19: [exec] curl --show-error --silent --include --output /dev/null --write-out '%{http_code}' --caCert ./ca.cert 'https://localhost:8443/core-data/api/v3/ping'
--- PASS: TestChangeTLSCert (0.40s)
```
GIthub Action:
```
exec.go:19: [exec] curl --show-error --silent --include --output /dev/null --write-out '%{http_code}' --caCert ./ca.cert 'https://localhost:8443/core-data/api/v3/ping'
    exec.go:69: [stdout] 000
    exec.go:72: [stderr] curl: (60) SSL certificate problem: self-signed certificate
        More details here: https://curl.se/docs/sslcerts.html
        
        curl failed to verify the legitimacy of the server and therefore could not
        establish a secure connection to it. To learn more about this situation and
        how to fix it, please visit the web page mentioned above.
    exec.go:75: exit status 60
--- FAIL: TestChangeTLSCert (0.33s)
```

File: test/suites/edgexfoundry/secrets_config_test.go [hunk 3]
```diff
@@ -41,3 +44,30 @@ func TestAddAPIGatewayUser(t *testing.T) {
...
+	createTLSCert(t)
+
+	t.Log("Calling API gateway using new TLS certificates:", coreDataPingEndpoint)
+	utils.WaitServiceOnline(t, 60, utils.ServicePort("nginx(https)"))
+
+	const caCertFile = "./ca.cert"
+	// Note: %%	is a literal percent sign
+	// Note: The path to the ca.cert file is created by create-tls-certificates.sh
+	code, _, _ := utils.Exec(t, fmt.Sprintf(
+		"curl --show-error --silent --include --output /dev/null --write-out '%%{http_code}' --caCert %s '%s'",
```
Comment (farshidtz) [inline test/suites/edgexfoundry/secrets_config_test.go:61]:
```suggestion
		"curl --show-error --silent --include --output /dev/null --write-out '%%{http_code}' --cacert %s '%s'",
```

Comment (farshidtz) [inline test/suites/edgexfoundry/secrets_config_test.go:61]:
```suggestion
	// Note: The ca.cert file is created by create-tls-certificates.sh
	code, _, _ := utils.Exec(t, fmt.Sprintf(
		"curl --verbose --show-error --silent --include --output /dev/null --write-out '%%{http_code}' --cacert %s '%s'",
```
Argument is wrong.
Verbose flag helps with debugging.

File: test/utils/create-tls-certificates.sh [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash -e
+
+SERVER_CERT_FILE=server.cert
+SERVER_KEY_FILE=server.key
+SERVER_CSR_FILE=server.csr
+CA_CERT_FILE=ca.cert
+CA_KEY_FILE=ca.key
```
Comment (farshidtz) [inline test/utils/create-tls-certificates.sh:7]: I suggest putting all these files in a directory to easily remove them during test cleanup.

File: test/utils/create-tls-certificates.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+openssl req -new -x509 -sha256 -key $CA_KEY_FILE -out $CA_CERT_FILE -subj "/CN=snap-testing-ca"
+# Generate the Server Certificate Private Key
+openssl ecparam -name prime256v1 -genkey -noout -out $SERVER_KEY_FILE
+# Generate the Server Certificate Signing Request
+openssl req -new -sha256 -key $SERVER_KEY_FILE -out $SERVER_CSR_FILE -subj "/CN=localhost"
+# Generate the Server Certificate
+openssl x509 -req -in $SERVER_CSR_FILE -CA $CA_CERT_FILE -CAkey $CA_KEY_FILE -CAcreateserial -out $SERVER_CERT_FILE -days 1000 -sha256
+
+# copy the files to a directory that the snap has permission to see
+sudo cp server.cert server.key /var/snap/edgexfoundry/common
```
Comment (farshidtz) [inline test/utils/create-tls-certificates.sh:21]: Use the environment variables for file names.
Move files so an explicit removal isn't needed on line 28.

File: test/suites/edgexfoundry/secrets_config_test.go [hunk 6]
```diff
@@ -41,3 +44,30 @@ func TestAddAPIGatewayUser(t *testing.T) {
 	require.NoError(t, err)
 	t.Logf("Output: %s", body)
 }
+
+// TestChangeTLSCert creats new TLS certificate and calls API gateway to verify the use of new certificate
+// https://docs.edgexfoundry.org/3.0/getting-started/Ch-GettingStartedSnapUsers/#changing-tls-certificates
+func TestChangeTLSCert(t *testing.T) {
+	t.Log("Create TLS certificate")
```
Comment (farshidtz) [inline test/suites/edgexfoundry/secrets_config_test.go:51]: Please remove certificate as part of testing cleanup.

**Review by farshidtz — CHANGES_REQUESTED** [review]

**Review by farshidtz — APPROVED** [review]
Looks good. Thanks.

Minor improvements that can be done in future in `test/utils/create-tls-certificates.sh`:
- Create temp directory only if not exited (instead of erroring out)
- Move certs to snap instead of copy and remove
- Add some logging to show where certs were created, perhaps enable xtrace instead?

**Pull request merged**